Add satisfiable tolerance parameter to TSR constructor

dqyi11 opened the pull request.

jslee02 commented:
I think we could simply pass the tolerance to `TSR::isSatisfied(...)`. How do you think @mkoval ?

dqyi11 commented:
@jslee02   isSatisfied() is inherited from a base class Testable. 
We will still have to add similar arguments in the constructors of adapter classes unless we change isSatisfiable() of Testable.

jslee02 commented:
> We will still have to add similar arguments in the constructors of adapter classes unless we change isSatisfiable() of Testable.

:+1: I thought other `Testable` classes also need some tolerance but it seems not.

jslee02 commented on include/aikido/constraint/TSR.hpp:
```diff
@@ -44,7 +44,8 @@ class TSR : public Sampleable,
           Eigen::Matrix<double, 6, 2>::Zero(),
-      const Eigen::Isometry3d& _Tw_e = Eigen::Isometry3d::Identity());
+      const Eigen::Isometry3d& _Tw_e = Eigen::Isometry3d::Identity(),
+      const double& _satisfiableTolerance = 1e-6);
```
Nit: Passing const reference for primitive type is pointless. Please change to `double`.

jslee02 commented on include/aikido/constraint/TSR.hpp:
```diff
@@ -57,7 +58,8 @@ class TSR : public Sampleable,
           Eigen::Matrix<double, 6, 2>::Zero(),
-      const Eigen::Isometry3d& _Tw_e = Eigen::Isometry3d::Identity());
+      const Eigen::Isometry3d& _Tw_e = Eigen::Isometry3d::Identity(),
+      const double& _satisfiableTolerance = 1e-6);
```
ditto

jslee02 commented on include/aikido/constraint/TSR.hpp:
```diff
@@ -124,9 +126,13 @@ class TSR : public Sampleable,
 private:
   std::unique_ptr<util::RNG> mRng;
   std::shared_ptr<statespace::SE3> mStateSpace;
+
```
Nit: Could we remove this empty line?

jslee02 commented on src/constraint/TSR.cpp:
```diff
@@ -79,25 +79,28 @@ class TSRSampleGenerator : public SampleGenerator
     , mTw_e(_Tw_e)
     , mRng(std::move(_rng))
     , mStateSpace(std::make_shared<SE3>())
+    , mSatisfiableTolerance(_satisfiableTolerance)
```
Initialize `mSatisfiableTolerance` before `mRng` and `mStateSpace` to resolve the [warning](https://travis-ci.org/personalrobotics/aikido/builds/223337806#L2840).

jslee02 commented on src/constraint/TSR.cpp:
```diff
@@ -79,25 +79,28 @@ class TSRSampleGenerator : public SampleGenerator
     , mRng(std::unique_ptr<util::RNG>(
           new util::RNGWrapper<std::default_random_engine>(0)))
     , mStateSpace(std::make_shared<SE3>())
+    , mSatisfiableTolerance(_satisfiableTolerance)
```
ditto

jslee02 commented on src/constraint/TSR.cpp:
```diff
@@ -109,6 +112,7 @@ TSR::TSR(const TSR& other)
     , mTw_e(other.mTw_e)
     , mRng(std::move(other.mRng->clone()))
     , mStateSpace(std::make_shared<SE3>())
+    , mSatisfiableTolerance(other.mSatisfiableTolerance)
```
ditto

jslee02 commented on src/constraint/TSR.cpp:
```diff
@@ -120,6 +124,7 @@ TSR::TSR(TSR&& other)
     , mTw_e(other.mTw_e)
     , mRng(std::move(other.mRng))
     , mStateSpace(std::make_shared<SE3>())
+    , mSatisfiableTolerance(other.mSatisfiableTolerance)
```
ditto

dqyi11 commented:
@jslee02  I addressed all your comments. Why does the order matter? 
Also I put it as a public variable, because I see other member variables are public. 
Shall I move it to private/protected and write get/set?

jslee02 commented on include/aikido/constraint/TSR.hpp:
```diff
@@ -44,7 +44,8 @@ class TSR : public Sampleable,
           Eigen::Matrix<double, 6, 2>::Zero(),
-      const Eigen::Isometry3d& _Tw_e = Eigen::Isometry3d::Identity());
+      const Eigen::Isometry3d& _Tw_e = Eigen::Isometry3d::Identity(),
+      const double _satisfiableTolerance = 1e-6);
```
Nit: `const` is meaningless for primitive types please change this to `double _satisfiableTolerance = 1e-6`

gilwoolee replied:
Need docstring for the new variable.

jslee02 commented on include/aikido/constraint/TSR.hpp:
```diff
@@ -57,7 +58,8 @@ class TSR : public Sampleable,
           Eigen::Matrix<double, 6, 2>::Zero(),
-      const Eigen::Isometry3d& _Tw_e = Eigen::Isometry3d::Identity());
+      const Eigen::Isometry3d& _Tw_e = Eigen::Isometry3d::Identity(),
+      const double _satisfiableTolerance = 1e-6);
```
ditto

jslee02 commented on include/aikido/constraint/TSR.hpp:
```diff
@@ -112,6 +114,10 @@ class TSR : public Sampleable,
 
+  double getSatisfiableTolerance();
+   
+  void setSatisfiableTolerance(const double satisfiableTolerance);
```
Nit: missing docstring

jslee02 commented on include/aikido/constraint/TSR.hpp:
```diff
@@ -112,6 +114,10 @@ class TSR : public Sampleable,
   bool project(const statespace::StateSpace::State* _s,
       statespace::StateSpace::State* _out) const override;
 
+  double getSatisfiableTolerance();
```
Nit: missing docstring

jslee02 commented on src/constraint/TSR.cpp:
```diff
@@ -79,10 +79,12 @@ class TSRSampleGenerator : public SampleGenerator
 TSR::TSR(std::unique_ptr<util::RNG> _rng, const Eigen::Isometry3d& _T0_w,
-         const Eigen::Matrix<double, 6, 2>& _Bw, const Eigen::Isometry3d& _Tw_e)
+         const Eigen::Matrix<double, 6, 2>& _Bw, const Eigen::Isometry3d& _Tw_e,
+         const double _satisfiableTolerance)
```
ditto

jslee02 commented on src/constraint/TSR.cpp:
```diff
@@ -91,10 +93,11 @@ TSR::TSR(std::unique_ptr<util::RNG> _rng, const Eigen::Isometry3d& _T0_w,
 //=============================================================================
 TSR::TSR(const Eigen::Isometry3d& _T0_w, const Eigen::Matrix<double, 6, 2>& _Bw,
-         const Eigen::Isometry3d& _Tw_e)
+         const Eigen::Isometry3d& _Tw_e, const double _satisfiableTolerance)
```
ditto

gilwoolee commented on include/aikido/constraint/TSR.hpp:
```diff
@@ -112,6 +114,14 @@ class TSR : public Sampleable,
 
+  /// Get the staisfiable tolerance for isSatisfiable.
+  /// \param[out] _out Satisfialble tolerance, double.
+  double getSatisfiableTolerance();
```
Change this to `getTestableTolerance()`. Although, I wonder why these set and get methods are necessary. Why do we want to change the tolerance once a TSR is created?

gilwoolee commented on include/aikido/constraint/TSR.hpp:
```diff
@@ -112,6 +114,14 @@ class TSR : public Sampleable,
+  double getSatisfiableTolerance();
+  
+  /// Set the staisfiable tolerance for isSatisfiable.
+  /// \param _satisfiableTolerance Satisfiable tolerance to set.
```
Typo. Though not sure why we need this method.

gilwoolee commented on include/aikido/constraint/TSR.hpp:
```diff
@@ -124,7 +134,10 @@ class TSR : public Sampleable,
+
 private:
+  /// Tolerance for checking satisfiability
+  double mSatisfiableTolerance;
```
Consider changing this to `Eigen::Vector6d`

dqyi11 replied:
Yes, I also think Eigen::Vector6d makes more sense.

dqyi11 commented:
I have addressed @gilwoolee 's comments.
I have not changed it to Vector6d, because the current implementation takes the norm of the distance vector to compare with a double threshold. I will make the change if comparing two vectors is expected.
Set() and Get() are still kept. Will keep or remove after hearing back from @mkoval  and @jslee02 .

mkoval commented on include/aikido/constraint/TSR.hpp:
```diff
@@ -44,7 +44,8 @@ class TSR : public Sampleable,
       const Eigen::Isometry3d& _T0_w = Eigen::Isometry3d::Identity(),
```
Add a docstring for `_testableTolerance `.

dqyi11 replied:
Fixed

mkoval commented on include/aikido/constraint/TSR.hpp:
```diff
@@ -44,7 +44,8 @@ class TSR : public Sampleable,
           Eigen::Matrix<double, 6, 2>::Zero(),
-      const Eigen::Isometry3d& _Tw_e = Eigen::Isometry3d::Identity());
+      const Eigen::Isometry3d& _Tw_e = Eigen::Isometry3d::Identity(),
+      const double _testableTolerance = 1e-6);
```
Don't mark arguments that are passed by value as `const` in function declarations.

dqyi11 replied:
Fixed

mkoval commented on include/aikido/constraint/TSR.hpp:
```diff
@@ -57,7 +58,8 @@ class TSR : public Sampleable,
           Eigen::Matrix<double, 6, 2>::Zero(),
-      const Eigen::Isometry3d& _Tw_e = Eigen::Isometry3d::Identity());
+      const Eigen::Isometry3d& _Tw_e = Eigen::Isometry3d::Identity(),
+      const double _testableTolerance = 1e-6);
```
Don't mark arguments that are passed by value as `const` in function declarations.

dqyi11 replied:
Fixed

mkoval commented on include/aikido/constraint/TSR.hpp:
```diff
@@ -57,7 +58,8 @@ class TSR : public Sampleable,
   TSR(const Eigen::Isometry3d& _T0_w = Eigen::Isometry3d::Identity(),
```
Add a docstring for `_testableTolerance `.

dqyi11 replied:
added.

mkoval commented on include/aikido/constraint/TSR.hpp:
```diff
@@ -112,6 +114,14 @@ class TSR : public Sampleable,
+  
+  /// Set the testable tolerance used in isSatisfiable.
+  /// \param _testableTolerance Testable tolerance to set.
+  void setTestableTolerance(const double _testableTolerance);
```
Don't mark arguments that are passed by value as `const` in function declarations.

dqyi11 replied:
Fixed

jslee02 commented:
As a third option, we could provide either way by providing `class TSR::ErrorMethod` that has a virtual function that measures the error and concrete classes that implement the error measurements. Then use a reasonable one as the default.

gilwoolee commented:
Based on what @mkoval said, `double` tolerance seems to be fine since it's only about _numerical precision_. @jslee02 's suggestion seems reasonable, but it may be an overcomplication if this is only for numerical precision.

I'm still not convinced that we need `setTolerance` and `getTolerance`.  Why do we want to change tolerance after initializing it? @mkoval @jslee02 thoughts?

jslee02 commented:
> Why do we want to change tolerance after initializing it? @mkoval @jslee02 thoughts?

I don't have a compelling reason to have the getters/setters but also not to have. One use case I can think of now is doing planning changing the error tolerance when it's failed to see if the failure is because of it. I'm fine either leaving them or not.
